monix-reactive/shared/src/main/scala/monix/reactive/Consumer.scala

```diff
@@ -369,6 +394,74 @@ object Consumer {
     }
   }
 
+  /** Implementation for [[Consumer.partitionParallel]] */
+  private final class PartitionParallelConsumer[A](nrPartitions: Int, overflowStrategy: OverflowStrategy.Synchronous[A], partition: A => Int, consumer: PartitionedConsumer[A]) extends Consumer[A, Unit] {
+    override def createSubscriber(cb: Callback[Unit], s: Scheduler): (Subscriber[A], AssignableCancelable) = {
+
+      val errors = scala.collection.mutable.ArrayBuffer.empty[Throwable]
+
+      val obs = PublishSubject[A]()
+      val pub = obs.whileBusyBuffer(overflowStrategy)
+      val isUpstreamDone = AtomicBoolean(false)
+
+      val refCount = RefCountCancelable { () =>
+        errors.synchronized {
+          if (errors.nonEmpty) {
+            cb.onError(CompositeException(errors))
+          } else {
+            cb.onSuccess(())
+          }
+        }
+      }
+
+      def subscriber(part: Int): Cancelable = {
+
+        def consumerSubscriber(refID: Cancelable) = new Subscriber[Seq[A]] {
+          implicit def scheduler: Scheduler = s
+
+          def onNext(elements: Seq[A]) =
+            consumer.consume(part, elements).coeval.value match {
+              case Left(future) =>
+                future.map(_ => Continue)
+              case Right(()) =>
+                Continue
+            }
+
+          def onError(ex: Throwable) = errors.synchronized {
+            errors += ex
+            onComplete()
+          }
+
+          def onComplete() = {
+            if (!isUpstreamDone.getAndSet(true)) {
+              refCount.cancel()
+            }
+
+            refID.cancel()
+          }
+        }
+
+        pub
+          .filter(x => partition(x) % nrPartitions == part)
+          .bufferIntrospective(s.executionModel.recommendedBatchSize)
+          .subscribe(consumerSubscriber(refCount.acquire()))
+      }
+
+      val out = new Subscriber[A] {
+        implicit def scheduler: Scheduler = s
+
+        def onNext(elem: A) = obs.onNext(elem)
+        def onError(ex: Throwable) = obs.onError(ex)
+        def onComplete() = obs.onComplete()
+      }
+
+      (0 until nrPartitions).foreach(subscriber)
+
+      out -> AssignableCancelable.dummy
```

Comment:
We've got a problem: in case all subscribers stop due to errors or something, they'll get unsubscribed for your `PublishSubject`, however the source does not stop and continues to stream events. It would be cool to stop the source as well.

This `AssignableCancelable` reference here can be used to cancel the source. You could return an actual `SingleAssignmentCancelable()` that you could then cancel in your logic for `RefCountCancelable` above.